### PR TITLE
EVG-6700: Don't allow DAG dispatcher return taskGroup items with unsatisfied dependencies 

### PR DESCRIPTION
**Pull request by StuartButler**

### CODE REVIEW

File: model/task_queue_service_dependency.go [hunk 1]
```diff
@@ -354,6 +345,27 @@ func (t *basicCachedDAGDispatcherImpl) nextTaskGroupTask(unit schedulableUnit) *
 			return nil
 		}
 
+		// Check if its dependencies have been met.
+		depsMet, err := nextTaskFromDB.DependenciesMet(make(map[string]task.Task))
+		if err != nil {
+			grip.Warning(message.Fields{
```
Comment (tychoish) [inline model/task_queue_service_dependency.go]: use message.WrapError(err, message.Fields{})

File: model/task_queue_service_dependency.go [hunk 2]
```diff
@@ -354,6 +345,27 @@ func (t *basicCachedDAGDispatcherImpl) nextTaskGroupTask(unit schedulableUnit) *
...
+				"dispatcher": "dependency-task-dispatcher",
+				"function":   "FindNextTask",
+				"message":    "error checking dependencies for task",
+				"outcome":    "skip and continue",
+				"task":       nextTaskFromDB.Id,
+				"error":      err.Error(),
```
Comment (tychoish) [inline model/task_queue_service_dependency.go]: omit this line.

File: model/task_queue_service_dependency.go [hunk 3]
```diff
@@ -354,6 +345,27 @@ func (t *basicCachedDAGDispatcherImpl) nextTaskGroupTask(unit schedulableUnit) *
 			return nil
 		}
 
+		// Check if its dependencies have been met.
+		depsMet, err := nextTaskFromDB.DependenciesMet(make(map[string]task.Task))
```
Comment (tychoish) [inline model/task_queue_service_dependency.go]: what's the point of passing an empty but constructed map to this function? is it a mechanism that we should be using to cache? should we make a variant of this method that doesn't take an argument? Something smells wrong about this.

File: model/task_queue_service_dependency.go [hunk 4]
```diff
@@ -175,17 +176,17 @@ func (t *basicCachedDAGDispatcherImpl) rebuild(items []TaskQueueItem) error {
...
+			// addEdge(A, B) means that B depends on A.
+			if err := t.addEdge(dependency, item.Id); err != nil {
 				return errors.Wrapf(err, "failed to create in-memory task queue of TaskQueueItems for distro '%s'; error defining a DirectedGraph incorporating task dependencies", t.distroID)
 			}
 		}
 	}
 
 	// Sort the graph. Use a lexical sort to resolve ambiguities, because node order is the order that we received these in.
-	sorted, err := topo.SortStabilized(t.graph, lexicalReversed)
+	sorted, err := topo.SortStabilized(t.graph, nil)
```
Comment (tychoish) [inline model/task_queue_service_dependency.go]: the comment is now out of date, and I'm not really sure I understand this change

Comment (StuartButler) [reply model/task_queue_service_dependency.go]: Removed.

File: model/task_queue_service_dependency.go (hunk 3 above)
Comment (StuartButler) [reply model/task_queue_service_dependency.go]: As discussed in person: will pass reference.

File: model/task_queue_service_dependency.go [hunk 5]
```diff
@@ -128,7 +129,7 @@ func (t *basicCachedDAGDispatcherImpl) addEdge(from string, to string) error {
 			"node_id":    fromNodeID,
 		})
 
-		return errors.New(fmt.Sprintf("cannot add a self edge to task '%s'", from))
+		return fmt.Errorf("cannot add a self edge to task '%s'", from)
```
[the remark this answers is not in the document]
Comment (StuartButler) [reply model/task_queue_service_dependency.go]: Done.

File: model/task_queue_service_dependency.go (hunk 1 above)
Comment (StuartButler) [reply model/task_queue_service_dependency.go]: Done.

File: model/task_queue_service_dependency.go (hunk 2 above)
Comment (StuartButler) [reply model/task_queue_service_dependency.go]: Done.

File: model/task_queue_service_dependency.go [hunk 6]
```diff
@@ -354,6 +342,27 @@ func (t *basicCachedDAGDispatcherImpl) nextTaskGroupTask(unit schedulableUnit) *
...
+				"function":   "nextTaskGroupTask",
+				"message":    "error checking dependencies for task",
+				"outcome":    "skip and continue",
+				"task":       nextTask.Id,
+			}))
+			continue
+		}
+
+		if !depsMet {
+			// TO DO do we need to set both of theses?
```
Comment (bsamek) [inline model/task_queue_service_dependency.go]: Could you check this before you merge?

Comment (StuartButler) [reply model/task_queue_service_dependency.go]: Sure - this is still a WIP.

Comment (StuartButler) [reply model/task_queue_service_dependency.go]: Both assignment statements give the same desired affect.

**Comment by bsamek** [comment]
@StuartButler Is this still a WIP?

**Pull request merged**